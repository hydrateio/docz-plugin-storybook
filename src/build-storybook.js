// TODO: rely on @storybook/react as a peer-dependency and probe its version at
// runtime for v3 vs v4 build process

import build from '@storybook/react/standalone'
import detect from 'detect-port'
import handler from 'serve-handler'
import http from 'http'
import pRetry from 'p-retry'
import puppeteer from 'puppeteer'
import readPkgUp from 'read-pkg-up'
import tempy from 'tempy'

module.exports = async (opts = { }) => {
  // build storybook standalone with custom webpack config
  // -----------------------------------------------------

  const { pkg } = await readPkgUp()
  const outputDir = tempy.directory()

  await build({
    mode: 'static',
    packageJson: pkg,
    configDir: './.storybook',
    outputDir,
    port: 60001,
    host: 'localhost',
    ...opts,
    frameworkPresets: [
      require.resolve('./storybook-preset.js'),
      ...(opts.frameworkPresets || [])
    ]
  })

  // console.log('storybook webpack success', outputDir)

  // serve static bundle
  // -------------------

  const server = http.createServer((request, response) => {
    return handler(request, response, {
      public: outputDir
    })
  })

  const port = await detect(49402)
  const baseUrl = `http://localhost:${port}`
  await new Promise((resolve, reject) => {
    server.listen(port, () => {
      resolve()
    })
  })

  // launch headless chrome and capture bundle's storybook
  // -----------------------------------------------------

  const browser = await puppeteer.launch()
  const page = await browser.newPage()

  await page.goto(`${baseUrl}/iframe.html`)
  const storybook = await pRetry(() => page.evaluate(() => {
    const globalHook = '__DOCZ_PLUGIN_STORYBOOK_CLIENT_API__'

    return window[globalHook].getStorybook()
  }))

  await Promise.all([
    browser.close(),
    new Promise((resolve, reject) => {
      server.close((err) => {
        if (err) reject(err)
        else resolve()
      })
    })
  ])

  return storybook
}
