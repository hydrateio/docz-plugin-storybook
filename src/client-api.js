import StoryStore from './story-store'

export default class ClientAPI {
  constructor({ storyStore = new StoryStore() } = {}) {
    this._storyStore = storyStore
    this._addons = {}
    this._globalParameters = {}
  }

  get store () {
    return this._storyStore
  }

  setAddon = (addon) => {
    this._addons = {
      ...this._addons,
      ...addon
    }
  }

  addDecorator = (decorator) => {
    // TODO: noop
  }

  addParameters = (parameters) => {
    this._globalParameters = parameters
  }

  clearDecorators = () => {
    // TODO: noop
  }

  storiesOf = (kind, m) => {
    if (!kind && typeof kind !== 'string') {
      throw new Error('Invalid or missing kind provided for stories, should be a string')
    }

    if (!m) {
      console.warn(
        `Missing 'module' parameter for story with a kind of '${kind}'. It will break your HMR`
      )
    }

    if (m && m.hot && m.hot.dispose) {
      m.hot.dispose(() => {
        this._storyStore.removeStoryKind(kind)
        this._storyStore.incrementRevision()
      })
    }

    let localParameters = {}
    const api = {
      kind
    }

    // apply addons
    Object.keys(this._addons).forEach(name => {
      const addon = this._addons[name]
      api[name] = (...args) => {
        addon.apply(api, args)
        return api
      }
    })

    api.add = (storyName, getStory, parameters) => {
      if (typeof storyName !== 'string') {
        throw new Error(`Invalid or missing storyName provided for a "${kind}" story.`)
      }

      if (this._storyStore.hasStory(kind, storyName)) {
        console.warn(`Story of "${kind}" named "${storyName}" already exists`)
      }

      const fileName = m ? m.id : null

      const allParam = { fileName }

      const paramScopes = [
        this._globalParameters,
        localParameters,
        parameters
      ]

      paramScopes.forEach((params) => {
        if (params) {
          Object.keys(params).forEach(key => {
            if (Array.isArray(params[key])) {
              allParam[key] = params[key]
            } else if (typeof params[key] === 'object') {
              allParam[key] = { ...allParam[key], ...params[key] }
            } else {
              allParam[key] = params[key]
            }
          })
        }
      })

      // Add the fully decorated getStory function.
      this._storyStore.addStory(
        kind,
        storyName,
        getStory,
        allParam
      )

      console.log('add', kind, storyName)
      return api
    }

    api.addDecorator = decorator => {
      // TODO: noop
      return api
    }

    api.addParameters = parameters => {
      // TODO: noop
      return api
    }

    return api
  }

  getStorybook = () => {
    return this._storyStore.getStoryKinds().map((kind) => {
      const fileName = this._storyStore.getStoryFileName(kind)

      const stories = this._storyStore.getStories(kind).map((name) => {
        const render = this._storyStore.getStoryWithContext(kind, name)
        return { name, render }
      })

      return { kind, fileName, stories }
    })
  }
}
