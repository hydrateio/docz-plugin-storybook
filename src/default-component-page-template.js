const defaultStorybookContentsRenderer = (kind) => '<StorybookContents />'

const defaultStorybookContentsImportsRenderer = (kindSlug) => `import StorybookContents from './${kindSlug}-storybook.mdx'`

export default ({ kind, stories, component, kindSlug, storybookContentsRenderer, storybookContentsImportsRenderer }) => {
  const routes = kind.split('/').map((p) => p.trim())

  let kindTitle = kind
  let menu = null

  if (routes.length > 1) {
    kindTitle = routes.slice(1).join(' - ')
    menu = routes[0]
  }

  const storybookRenderer = storybookContentsRenderer || defaultStorybookContentsRenderer
  const storybookImportsRenderer = storybookContentsImportsRenderer || defaultStorybookContentsImportsRenderer

  return `
---
name: ${kindTitle}
${menu ? `menu: ${menu}` : ''}
---

import { Playground } from 'docz'
${storybookImportsRenderer(kindSlug, component)}

# ${kind}

${storybookRenderer(kind, component)}
`
}
