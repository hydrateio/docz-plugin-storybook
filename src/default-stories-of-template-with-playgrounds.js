const renderComponentPlayground = (component) => {
  if (!component) return ''

  const { children, ...defaultPlaygroundProps } = component.defaultPlaygroundProps || {}

  const propString = Object.keys(defaultPlaygroundProps).map(key => {
    return `${key}={${defaultPlaygroundProps[key]}}`
  }).join(' ')

  if (children) {
    return `<Playground>
  <${component.importName} ${propString}>
    ${children}
  </${component.importName}>
</Playground>
`
  }

  return `<div>&lt;Playground&gt;
  &lt;${component.importName} ${propString} /&gt;
&lt;Playground&gt;</div>
`
}

const renderComponentImports = (component) => {
  if (!component) return ''

  const imports = [
    { importName: component.importName, importPath: component.importPath },
    ...component.otherImports
  ]

  return imports.map(imp => `import ${imp.importName} from '${imp.importPath}'`).join('\n')
}

export default ({ kind, stories, component }) => {
  return `
---
routable: false
---

import { Story } from 'docz-plugin-storybook/dist/react'
import { Playground } from 'docz'
${renderComponentImports(component)}

${stories.map(({ name }) => `## ${name}

<Story kind="${kind}" name="${name}" />

${renderComponentPlayground(component)}
`).join('\n')}
`
}
