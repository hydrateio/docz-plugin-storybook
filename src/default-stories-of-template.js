export default ({ kind, stories, component, defaultPlaygroundProps }) => {
  return `
---
routable: false
---

import { Story } from 'docz-plugin-storybook/dist/react'

${stories.map(({ name }) => `## ${name}

<Story kind="${kind}" name="${name}" />
`).join('\n')}
`
}
