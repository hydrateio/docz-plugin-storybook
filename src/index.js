import { createPlugin } from 'docz-core'
import * as path from 'path'
import fs from 'fs-extra'
import paths from './paths'
import buildStorybook from './build-storybook'

export const storybook = (opts = { }) => {
  const {
    // these options are intended to mimic the storybook cli options
    configDir = paths.storybook.configDir,
    storyWrapper,
    manual = false
  } = opts

  const storybookConfigPath = path.resolve(configDir, paths.storybook.config)
  const storybookFiles = []

  return createPlugin({
    setConfigPromise: async (config) => {
      if (!manual) {
        const storybook = await buildStorybook({ configDir })
        console.log('storybook', JSON.stringify(storybook, null, 2))

        await fs.ensureDir(paths.temp)
        await Promise.all(storybook.map((storyKind) => {
          const { kind, stories } = storyKind

          const content = `
---
name: ${kind} Stories
---

import { Story } from 'docz-plugin-storybook/dist/react'

# ${kind}

${stories.map(({ name }) => `
## ${name}

<Story kind='${kind}' name='${name}' />

`)}
`
          const file = path.join(paths.temp, `${kind}-storybook.mdx`)
          storybookFiles.push(file)
          return fs.writeFile(file, content)
        }))
      }

      // config.files = ensureArray(config.files).concat([
      //   path.join(paths.temp, '*.mdx')
      // ])

      console.log('docz config')
      console.log('-'.repeat(80))
      console.log(JSON.stringify(config, null, 2))
      console.log()
      console.log()

      return config
    },

    modifyFiles: (files) => {
      return files.concat(storybookFiles)
    },

    modifyBundlerConfig: (config, dev, args) => {
      config.entry = config.entry || {}
      config.entry.app = config.entry.app || []
      config.entry.app.push(storybookConfigPath)

      config.resolve = config.resolve || {}
      config.resolve.alias = config.resolve.alias || {}
      config.resolve.alias['@storybook/react'] = path.resolve(__dirname, './shim')

      // create a wrapper around each component for isolation and surface it as docz-plugin-storybook/story-wrapper
      config.resolve.alias['docz-plugin-storybook/story-wrapper'] = require.resolve(storyWrapper || './default-wrapper')

      console.log('webpack')
      console.log('-'.repeat(80))
      console.log(JSON.stringify(config, null, 2))
      console.log()
      console.log()

      return config
    }
  })
}
