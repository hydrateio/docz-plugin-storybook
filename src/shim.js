import ClientAPI from './client-api'

const clientAPI = new ClientAPI()

// mock the minimal set of exports from `@storybook/react`
export const storiesOf = clientAPI.storiesOf
export const setAddon = clientAPI.setAddon
export const addDecorator = clientAPI.addDecorator
export const addParameters = clientAPI.addParameters
export const clearDecorators = clientAPI.clearDecorators
export const getStorybook = clientAPI.getStorybook

export const forceReRender = () => { } // TODO: noop

export const configure = (loaders, module) => {
  if (loaders) {
    // load all stories into the client API via one or more imperative commonjs
    // 'require' trees
    // TODO: is this setTimeout necessary?
    setTimeout(() => loaders())
  }
}

// additional, private exports
export const _clientAPI = clientAPI
