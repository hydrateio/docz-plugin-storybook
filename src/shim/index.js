import ClientAPI from './client-api'
import './channel'

// TODO: possibly remove this from shim
// import './channel'

const globalHook = '__DOCZ_PLUGIN_STORYBOOK_CLIENT_API__'
const clientAPI = window[globalHook] || new ClientAPI()
window[globalHook] = clientAPI

// mock the minimal set of exports from `@storybook/react`
export const storiesOf = clientAPI.storiesOf
export const setAddon = clientAPI.setAddon
export const addDecorator = clientAPI.addDecorator
export const addParameters = clientAPI.addParameters
export const clearDecorators = clientAPI.clearDecorators
export const getStorybook = clientAPI.getStorybook

export const forceReRender = () => { } // TODO: noop

export const configure = (loaders, module) => {
  console.log('configuring storybook', loaders)
  if (loaders) {
    // load all stories into the client API via one or more imperative commonjs
    // 'require' trees
    loaders()
  }
}

// additional, private exports
export const _clientAPI = clientAPI
