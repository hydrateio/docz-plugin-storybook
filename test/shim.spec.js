import renderer from 'react-test-renderer'
import { getStorybook } from '../src/shim'

describe('Storybook Shim', () => {
  describe('against Basic Fixture', () => {
    let ButtonStory, story1, story2

    beforeAll(() => {
      require('./fixtures/basic')
      const storyBook = getStorybook()
      ButtonStory = storyBook[0]
      story1 = ButtonStory.stories[0]
      story2 = ButtonStory.stories[1]
    })

    it('should transform the storybook to an object', () => {
      expect(ButtonStory.kind).toBe('Button')
      expect(Array.isArray(ButtonStory.stories)).toBe(true)
      expect(ButtonStory.stories.length).toBe(2)
      expect(story1.name).toBe('with text')
      expect(story2.name).toBe('with some emoji')
    })

    it('should match story snapshots', () => {
      const Story1 = renderer.create(story1.render()).toJSON()
      const Story2 = renderer.create(story2.render()).toJSON()

      expect(Story1).toMatchSnapshot()
      expect(Story2).toMatchSnapshot()
    })
  })
})
