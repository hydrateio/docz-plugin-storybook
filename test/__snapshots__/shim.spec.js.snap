// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Storybook Shim against Basic Fixture should match story snapshots 1`] = `
<button>
  Hello Button
</button>
`;

exports[`Storybook Shim against Basic Fixture should match story snapshots 2`] = `
<button>
  <span
    aria-label="so cool"
    role="img"
  >
    😀 😎 👍 💯
  </span>
</button>
`;
